### PR TITLE
pipelined: Config ARP cache for upstream APN vlan

### LINKED ISSUE #2323: Add support for handling multiple upstream GW in PipelineD.
This involves probing Internet GW on each APN network. DHCP client need to use vlan tag associated APN to probe the GW node.

### PR DESCRIPTION
**Pull request by pshelar**
## Summary

Following patch uses mobilityD GWInfo API to get Internet GW Ip address.
The it periodically probe internet GW mac address using ARP. This way pipelineD
can get MAC address of upstream GW.

Signed-off-by: Pravin B Shelar <pbshelar@fb.com>

### CODE REVIEW

**Comment by pshelar** [comment]
This is stacked Pull request. For now please refer to commit "pipelined: Config ARP cache for upstream APN vlan" for changes related to this task.

File: lte/gateway/python/magma/pipelined/tests/test_inout_non_nat.py [hunk 1]
```diff
@@ -120,30 +162,95 @@ def setUp(self):
...
+        mocked_set_mobilityd_gw_info(IPAddress(address=ip_addr.packed,
+                                               version=IPBlock.IPV4),
+                                     "", vlan1)
 
+        ip_addr = ipaddress.ip_address("10.200.22.211")
+        mocked_set_mobilityd_gw_info(IPAddress(address=ip_addr.packed,
+                                               version=IPBlock.IPV4),
+                                     "", vlan2)
+
+        time.sleep(2)
```
Comment (koolzz) [inline lte/gateway/python/magma/pipelined/tests/test_inout_non_nat.py]: probably better to use hub.sleep? import ryu.hub

I recall using time sleep might have some ryu collisions

File: lte/gateway/python/magma/pipelined/app/inout.py [hunk 2]
```diff
@@ -171,7 +170,7 @@ def _install_default_middle_flows(self, dp):
                                   [], priority=flows.UE_FLOW_PRIORITY,
                                   output_port=self.config.mtr_port)
 
-    def _install_default_egress_flows(self, dp, mac_addr: str = ""):
+    def _install_default_egress_flows(self, dp, mac_addr: str = "", vlan: str = ""):
         """
         Egress table is the last table that a packet touches in the pipeline.
         Output downlink traffic to gtp port, uplink trafic to LOCAL
```
Comment (koolzz) [inline lte/gateway/python/magma/pipelined/app/inout.py:176]: Maybe add a short explanation of the default flows being installed

**Review by koolzz — APPROVED** [review]

File: lte/gateway/python/magma/pipelined/app/inout.py [hunk 3]
```diff
@@ -183,20 +185,43 @@ def _install_default_egress_flows(self, dp, mac_addr: str = ""):
...
+            uplink_match = MagmaMatch(direction=Direction.OUT,
+                                      vlan_vid=(vid, vid))
+        else:
+            uplink_match = MagmaMatch(direction=Direction.OUT)
+
         actions = []
         # avoid resetting mac address on switch connect event.
         if mac_addr == "":
-            mac_addr = self._current_upstream_mac
+            if vlan in self._current_upstream_mac_map:
```
Comment (ardzoht) [inline lte/gateway/python/magma/pipelined/app/inout.py]: nit: you can remove the vlan in map check and just do `mac_addr = self._current_upstream_mac_map.get(vlan, "")`

**Review by ardzoht — APPROVED** [review]

**Pull request merged**